Fix and enable sam4e ethernet port

nandojve opened the pull request:
This PR add all missing pieces to fix and enable SAM4E Ethernet port.

The changes were tested on SAM4E and SAMV71 boards using samples/net/telnet. Tests are conducted using DHCPv4, collecting stats and ping 8.8.8.8

nandojve committed:
soc: arm: atmel: Add sam4e gmac pins

Add sam4e gmac pin definitions to use ethernet driver.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: Kconfig.sam_gmac: Fix NOCACHE dependency

The NOCACHE_MEMORY can be enabled only for those MCU that support data
cache. The currently SoC that doesn't have data cache is SAM4E.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: sam_gmac: Add data cache checks

Improve data cache conditional build. Now data cache code is build
only if device have support to it. This enables GMAC driver for use
with devices that don't have data cache instructions.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: sam_gmac: Fix phy layer setup

The current setup of physical layer forces RMII interface. The code was
refactored to have only one point to select proper phy interface. Now,
GMAC driver works with both RMII or MII interface. The phy connection
type is now selected by device tree. The default phy connection is RMII.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: arm: sam4e_xpro: Add GMAC defaults

This enable GMAC driver by default if the end application selects
NETWORKING. The driver will set MII interface since is the only one
available for SAM4E and will select a random MAC value.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: eth: eth_sam_gmac: Split DMA queue flags

This cleans up DMA flags by separating the necessary flags for devices
with one or multiple RX/TX queues.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

stephanosio reviewed (commented):
We should add `supported: netif:eth` in `sam4e_xpro.yaml` too.

stephanosio commented:
Just a reminder that, if #23045 gets merged before this PR, we need to add device tree stuff here as well.

I already have the commits for that here:
https://github.com/stephanosio/zephyr/commit/a14216636747bf28107f7a7183ba75fec014752e
https://github.com/stephanosio/zephyr/commit/da8a86872adb1404363644c1138c34385acaf02b

nandojve committed:
dts: arm: atmel: Add SAM 4E GMAC instance

This commit adds a GMAC instance to the SAM 4E device tree.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: sam4e_xpro: Enable GMAC in device tree

This commit enables the newly added GMAC instance in the `sam4e_xpro`
device tree.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
> Just a reminder that, if #23045 gets merged before this PR, we need to add device tree stuff here as well.
> 
> I already have the commits for that here:
> [stephanosio@a142166](https://github.com/stephanosio/zephyr/commit/a14216636747bf28107f7a7183ba75fec014752e)
> [stephanosio@da8a868](https://github.com/stephanosio/zephyr/commit/da8a86872adb1404363644c1138c34385acaf02b)

nandojve committed:
drivers: eth: sam_gmac: Use DT max_frame_size option

This commit updates the Atmel SAM GMAC driver to select max frame size
value from the device tree. Now GMAC driver can operate with the three
different frame size options available.

The current supported values are: 1518, 1536 and 10240.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
dts: bindings: eth: Add GMAC options

This add the following options to GMAC device tree bindings:
 - max-frame-size
 - max-speed
 - phy-connection-type

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
dts: arm: atmel: Add SAM4E GMAC phy-connection-type

The SAM4E GMAC version can use only MII as phy-connection-type. This
update the current default RMII value to MII.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
@mnkp

nandojve committed:
drivers: ethernet: eth_sam_gmac.c: Fix macro name

Fix macro name from disable_all_queue_interrupt to
disable_all_priority_queue_interrupt.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

jukkar reviewed (approved):
LGTM

stephanosio commented:
#23045 has been merged. This PR can be rebased now.

nandojve committed:
boards: arm: sam_e70_xplained: Fix dts model and compatible

Set model according with SoC revision A or B. Compatible follows
same approach.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
boards: arm: sam_v71_xult: Fix dts model and compatible

Set model according with SoC revision A or B. Compatible follows
same approach.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
Rebased.

nandojve commented:
Rebase and fixed flash_sam 9d6530c

stephanosio commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -15,3 +16,32 @@ properties:
       required: true
       description: Number of hardware TX and RX queues
+
+    local-mac-address:
+      type: uint8-array
+      description:
+        Specifies the MAC address that was assigned to the network device
```
`local-mac-address` is already defined in `ethernet.yaml`, so no need to re-define it here; though the property description there could probably use some improvement.

nandojve replied:
indeed, fixed!

stephanosio commented on dts/bindings/ethernet/atmel,sam-gmac.yaml:
```diff
@@ -15,3 +16,32 @@ properties:
+        description:
+          Operation mode of the PHY interface
+        enum:
+          - rmii
+          - mii
```
Shouldn't this be:
```
- "rmii"
- "mii"
```
(cc @galak @ulfalizer)

nandojve replied:
fixed!

nvlsianpu reviewed (approved):
approve to the flash driver patch

nandojve commented:
rebase - flash fix was already merged.

vlotech commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -6,7 +6,7 @@
 	bool "Atmel SAM Ethernet driver"
 	depends on SOC_FAMILY_SAM
-	select NOCACHE_MEMORY
+	select NOCACHE_MEMORY if !SOC_SERIES_SAM4E
```
instead of using SAM4E, I would advise to use:
select NOCACHE_MEMORY if ARCH_HAS_NOCACHE_MEMORY_SUPPORT

nandojve replied:
Done!

nandojve commented:
rebase and revert dt_node_int_prop_int to Kconfig 694cd62 until a better solution. In all cases, path isn't a valid solution because SoCs can have different address space for same peripheral.

stephanosio commented on drivers/ethernet/Kconfig.sam_gmac:
```diff
@@ -15,7 +15,12 @@ if ETH_SAM_GMAC
 	int "Number of active hardware TX and RX queues"
 	default 1
-	range 1 $(dt_node_int_prop_int,/soc/ethernet@40050088,num-queues)
+	range 1 1 if SOC_SERIES_SAM4E
+	range 1 3 if !SOC_SERIES_SAM4E && \
+		     !SOC_ATMEL_SAME70_REVB && \
+		     !SOC_ATMEL_SAMV71_REVB
+	range 1 6 if SOC_ATMEL_SAME70_REVB || \
+		     SOC_ATMEL_SAMV71_REVB
```
TODO: Investigate how to deal with this (cc @galak)
https://github.com/zephyrproject-rtos/zephyr/pull/24057#discussion_r406572496

stephanosio replied:
@nandojve See #24256

vlotech committed:
drivers: eth: Kconfig.sam_gmac: Add sam4e support for queues

Fixed the ETH_SAM_GMAC_QUEUES config by adding if-statements per SoC
series.

Signed-off-by: Vincent van der Locht <vincent@vlotech.nl>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

vlotech committed:
include: linker: section_tags.h: Add __nocache fallback

Some drivers support NOCACHE_MEMORY sections. To have a default
fallback for systems where this isn't available or disabled a
fallback define is added.

Signed-off-by: Vincent van der Locht <vincent@vlotech.nl>
Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
rebase and added:

9ad5426 better approach to handle __nocache
62cca3b better way to handle in Kconfig

Thank you @stephanosio @galak @vlotech for all effort spend with the Kconfig issue.

nandojve commented:
Hi everyone, I was wondering if there is anything missing here? This is an important feature that have been blocked since Feb and will be great have it on Zephyr 2.3.
